### PR TITLE
PostgreSQL - Simple Composer for reduced memory usage

### PR DESCRIPTION
**Pull request by krnaveen14**
### Database name
PostgreSQL

# Pull request description
Simple Composer for reduced memory usage on instances with `> 100k` files

- Does not track metadata of files backed up (Files and TarFileSets)
- Does not support `WALG_DELTA_MAX_STEPS`, `WALG_DELTA_ORIGIN`, `WALG_USE_REVERSE_UNPACK`, `WALG_SKIP_REDUNDANT_TARS`, `catchup-push`

### Describe what this PR fix
Similar to Issue #738 , WAL-G consumes huge memory on backing up instance with millions of files. This is due to tracking the metadata of each file, maintaining the list of files in each TarFileSet and finally marshalling the data structure to text for creating `sentinel.json` file. Even though PR #740 tries to address the marshalling part by streaming / pipe it, significant memory usage on `BackupSentinelDto` data structure still persists.

Based on the `BackupSentinelDto` usage, it is needed only for configurations such as `DELTA`, `INCREMENTAL`, `RATING_COMPOSER`, `catchup-push` etc... Since we don't use any of the above mentioned configs, `Files` and `TarFileSets` data in `BackupSentinelDto` simply unused.

To reduce the memory usage of WAL-G on instances with millions of files and to make it available & configurable for the community as whole, `SimpleTarBallComposer` is implemented which does not track any of the file metadata, list of files in each TarFileSet and the list of TarFileSets itself.

Simple Composer is supported on both local and remote backups.

**Usage :** `wal-g backup-push [path] --simple-composer` or `WALG_USE_SIMPLE_COMPOSER=true` environment variable

### Please provide steps to reproduce
Create 500+ Databases with 1000+ Tables in each Database (not necessary to insert any data). Execute WAL-G `backup-push` to see the memory usage of `wal-g` gradually increases over time and failing with OOM after exhausting available memory (10GB in our case).

### CODE REVIEW

**Comment by krnaveen14** [comment]
@usernamedt waiting on approval for test execution and expecting code review suggestions

File: docs/PostgreSQL.md
Comment (usernamedt) [inline docs/PostgreSQL.md:239]: You can add an explicit note that simple mode does not support delta and catchup backups.

Also, currently, nothing prevents the user from making a delta or catchup backup with simple mode enabled, without any noticeable error. WAL-G should fail in this case and forbid creating a delta/catchup backup when the simple mode is enabled.

File: internal/databases/postgres/backup_push_handler.go
Comment (usernamedt) [inline internal/databases/postgres/backup_push_handler.go:104]: Using global variables should be avoided, pass this setting directly to the `NewTarballStreamer`

File: internal/databases/postgres/simple_tar_ball_composer.go
Comment (usernamedt) [inline internal/databases/postgres/simple_tar_ball_composer.go:1]: I don't see any need for the additional `TarBallComposer` here. `SimpleTarBall` composer does not introduce any new way to pack tarballs, just alters the `BundleFiles` behavior.

So I suggest removing the new composer and passing the SimpleBundleFiles to the regular composer. 

Just replace the `*RegularBundleFiles` with the `BundleFiles` interface [here](https://github.com/wal-g/wal-g/blob/3a5120f1cc208a69db6a834c979a4c97a4d2e7bc/internal/databases/postgres/regular_tar_ball_composer.go#L18) so the regular tarball composer could support both `SimpleBundleFiles` and `RegularBundleFiles`.

Comment (usernamedt) [reply internal/databases/postgres/simple_tar_ball_composer.go:1]: As for `TarFileSets`, you can do the similar:
1. Create the `TarFileSets` interface
2. Create `SimpleTarFileSets`/`NopTarFileSets` (new behavior) and `RegularTarFileSets` (old behavior)

File: internal/databases/postgres/tar_ball_composer.go
Comment (usernamedt) [inline internal/databases/postgres/tar_ball_composer.go:14]: With the `SimpleTarFileSets` implemented, this will be unnecessary.

**Review by usernamedt — CHANGES_REQUESTED** [review]
Hi!

This is definitely a cool feature, but it needs some adjustments.

**Review by usernamedt — COMMENTED** [review]
Hi!

Looks like this unit test is failing:

```
2021-08-17T11:25:48.0366780Z === RUN   TestGetFilesToUnwrap_UtilityFiles
2021-08-17T11:25:48.0367313Z     backup_test.go:49: 
2021-08-17T11:25:48.0367713Z         	Error Trace:	backup_test.go:49
2021-08-17T11:25:48.0378206Z         	Error:      	Not equal: 
2021-08-17T11:25:48.0378808Z         	            	expected: map[string]bool{"/global/pg_control":true, "backup_label":true, "tablespace_map":true}
2021-08-17T11:25:48.0379442Z         	            	actual  : map[string]bool(nil)
2021-08-17T11:25:48.0379793Z         	            	
2021-08-17T11:25:48.0380100Z         	            	Diff:
2021-08-17T11:25:48.0380664Z         	            	--- Expected
2021-08-17T11:25:48.0381016Z         	            	+++ Actual
2021-08-17T11:25:48.0381447Z         	            	@@ -1,6 +1,2 @@
2021-08-17T11:25:48.0381949Z         	            	-(map[string]bool) (len=3) {
2021-08-17T11:25:48.0382567Z         	            	- (string) (len=18) "/global/pg_control": (bool) true,
2021-08-17T11:25:48.0383204Z         	            	- (string) (len=12) "backup_label": (bool) true,
2021-08-17T11:25:48.0383852Z         	            	- (string) (len=14) "tablespace_map": (bool) true
2021-08-17T11:25:48.0384349Z         	            	-}
2021-08-17T11:25:48.0384701Z         	            	+(map[string]bool) <nil>
2021-08-17T11:25:48.0385035Z         	            	 
2021-08-17T11:25:48.0385506Z         	Test:       	TestGetFilesToUnwrap_UtilityFiles
2021-08-17T11:25:48.0386267Z --- FAIL: TestGetFilesToUnwrap_UtilityFiles (0.00s)
2021-08-17T11:25:48.0386909Z === RUN   TestGetFilesToUnwrap_NoMoreFiles
2021-08-17T11:25:48.0387661Z --- PASS: TestGetFilesToUnwrap_NoMoreFiles (0.00s)
```

File: internal/databases/postgres/bundle_files.go
Comment (usernamedt) [inline internal/databases/postgres/bundle_files.go:100]: `NopBundleFiles`?

File: cmd/pg/backup_push.go
Comment (usernamedt) [inline cmd/pg/backup_push.go:58]: Well, now this is not actually a composer type, but a flag to enable the low memory usage. So I suggest to rename  this to `--reduce-memory-usage` or something like this since it is not a new composer type.

**Review by usernamedt — CHANGES_REQUESTED** [review]
Seems OK, but need to change the flag

**Comment by krnaveen14** [comment]
Closing this in favor new PR #1101

**Pull request closed without merging**
